src/featbin/wav-reverberate.cc

```diff
@@ -0,0 +1,325 @@
+    BaseFloat samp_freq_input = input_wave.SampFreq();  // read sampling fequency
+    int32 num_samp_input = input_matrix.NumCols(),  // number of samples in recording
+          num_input_channel = input_matrix.NumRows();  // number of channels in recording
+    KALDI_VLOG(1) << "sampling frequency of input: " << samp_freq_input
+                  << " #samples: " << num_samp_input
+                  << " #channel: " << num_input_channel;
+    KALDI_ASSERT(input_channel < num_input_channel);
+
+    WaveData rir_wave;
+    {
+      Input ki(rir_file);
+      rir_wave.Read(ki.Stream());
+    }
+    const Matrix<BaseFloat> &rir_matrix = rir_wave.Data();
+    BaseFloat samp_freq_rir = rir_wave.SampFreq();
+    int32 num_samp_rir = rir_matrix.NumCols(),
+          num_rir_channel = rir_matrix.NumRows();
+    KALDI_VLOG(1) << "sampling frequency of rir: " << samp_freq_rir
+                  << " #samples: " << num_samp_rir
+                  << " #channel: " << num_rir_channel;
+    if (!multi_channel_output) {
+      KALDI_ASSERT(rir_channel < num_rir_channel);
+    }
+
+    Matrix<BaseFloat> noise_matrix;
+    if (!noise_file.empty()) {
+      WaveData noise_wave;
+      {
+        Input ki(noise_file);
+        noise_wave.Read(ki.Stream());
+      }
+      noise_matrix = noise_wave.Data();
+      BaseFloat samp_freq_noise = noise_wave.SampFreq();
+      int32 num_samp_noise = noise_matrix.NumCols(),
+            num_noise_channel = noise_matrix.NumRows();
+      KALDI_VLOG(1) << "sampling frequency of noise: " << samp_freq_noise
+                    << " #samples: " << num_samp_noise
+                    << " #channel: " << num_noise_channel;
+      if (multi_channel_output) {
+        KALDI_ASSERT(num_rir_channel == num_noise_channel);
+      } else {
+        KALDI_ASSERT(noise_channel < num_noise_channel);
+      }
+    }
+
+    int32 num_output_channel = (multi_channel_output ? num_rir_channel : 1);
+    Matrix<BaseFloat> out_matrix(num_output_channel, num_samp_input);
+
+    for (int32 output_channel = 0; output_channel < num_output_channel; output_channel++) {
+      Vector<BaseFloat> input(num_samp_input);
+      input.CopyRowFromMat(input_matrix, input_channel);
+
+      int32 this_rir_channel = (multi_channel_output ? output_channel : rir_channel);
+      Vector<BaseFloat> rir(num_samp_rir);
+      rir.CopyRowFromMat(rir_matrix, this_rir_channel);
+      rir.Scale(1 / MaxAbsolute(rir));
+
+      Vector<BaseFloat> noise(0);
+      if (!noise_file.empty()) {
+        noise.Resize(noise_matrix.NumCols());
+        int32 this_noise_channel = (multi_channel_output ? output_channel : noise_channel);
+        noise.CopyRowFromMat(noise_matrix, this_noise_channel);
+      }
+
+      DoReverberation(rir, &noise, &input, samp_freq_rir, snr_db, test_mode);
+
+      float input_max = MaxAbsolute(input);
+      for (int32 j = 0; j < input.Dim(); j++)
+        input(j) = (0.99 * input(j) / input_max) * pow(2, 15);
```

Comment:
On Aug 21, 2015 7:25 PM, "Daniel Povey" <notifications@github.com> wrote:
>
> In src/featbin/wav-reverberate.cc:
>
> > +      Vector<BaseFloat> rir(num_samp_rir);
> > +      rir.CopyRowFromMat(rir_matrix, this_rir_channel);
> > +      rir.Scale(1 / MaxAbsolute(rir));
> > +
> > +      Vector<BaseFloat> noise(0);
> > +      if (!noise_file.empty()) {
> > +        noise.Resize(noise_matrix.NumCols());
> > +        int32 this_noise_channel = (multi_channel_output ?
output_channel : noise_channel);
> > +        noise.CopyRowFromMat(noise_matrix, this_noise_channel);
> > +      }
> > +
> > +      DoReverberation(rir, &noise, &input, samp_freq_rir, snr_db,
test_mode);
> > +
> > +      float input_max = MaxAbsolute(input);
> > +      for (int32 j = 0; j < input.Dim(); j++)
> > +        input(j) = (0.99 * input(j) / input_max) * pow(2, 15);
>
> Guys,
> The solutions you are suggesting for the scaling problem are starting to
sound extremely complicated and hard to implement. I think what would be
best is to just state that this program normalizes the output so that the
overall signal energy after reverberation and adding noise will be the same
as the signal input. Then if you don't like the clipping, you scale down
the input. The code in wav-reverberate could, after scaling, clip the
signal to the range [-1, 1], or have the WaveDataNormalized::Write function
do this (and in either case, print out a warning). Naturally, class
WaveDataNormalized would have to document whether it expects the input to
be clipped, or whether it does the clipping itself.

OK. I am worried that after noise addition to have 0dB SNR, the output
signal+noise energy would be comparatively higher. Making its energy
similar to input might lead to a waveform with lower signal (not signal +
noise) energy than the input. This is the reason I suggest we just give the
user the option to specify a constant scaling factor. We can delegate the
responsibility of determining this factor to the user. The binary need not
provide an option to do this.

> Regarding Kirill's question about where code of this nature should go--
yes, in src/feat/. E.g. in src/feat/signal.h.
> Do either of Vijay or Kirill have time to make the changes to the
wave-reading code, and add the normalized versions? That should probably be
finalized before this pull request is finished.

I can do this. But I am not familiar with this module so I might take a day.
>
> —
> Reply to this email directly or view it on GitHub.